clustering_code/edchan-kmeans.py

```diff
@@ -0,0 +1,913 @@
+                    ylabel='Longitude',
+                    bands='band', 
+                    data_aspect=1,
+                    title=title,
+                    crs=ccrs.UTM(16))
+rgb_plot
+
+# %%
+fig, ax = plt.subplots()
+
+rgb_image = np.stack([mean_red, mean_green, mean_blue], axis=-1)
+rgb_image = (rgb_image - np.min(rgb_image)) / (np.max(rgb_image) - np.min(rgb_image))
+
+ax.imshow(rgb_image)
+title = f'Manuel Canal-Spanish Lake: Composite Image'
+ax.set_title(title)
+ax.set_aspect('equal')
+ax.set_xlabel('')
+ax.set_ylabel('')
+ax.set_xticks([])
+ax.set_yticks([])
+plt.show()
+
+# %% [markdown]
+# ## STEP 4: K-MEANS
+
+# %% [markdown]
+# Cluster your data by spectral signature using the k-means algorithm. 
+# 
+# **YOUR TASK:**
+# 1. Convert your DataArray into a **tidy** DataFrame of reflectance values (hint: check out the `.unstack()` method)
+# 2. Filter out all rows with no data (all 0s or any N/A values)
+# 3. Fit a k-means model. You can experiment with the number of groups to find what works best.
+
+# %%
+k_means_df = (
+    rgb_stacked
+    .drop('spatial_ref')
+    .to_dataframe(name='reflectance')
+    .unstack('band')
+)
+k_means_df
+
+# %%
+processed_xr = rgb_stacked
+# Running the fit and predict functions at the same time.
+# We can do this since we don't have target data.
+prediction = KMeans(n_clusters=3).fit_predict(k_means_df.values)
+
+# Reshaping the prediction array to match the original raster shape
+prediction = prediction.reshape(processed_xr['band'==4].shape)
+prediction
+
+# %% [markdown]
+# ## STEP 5: PLOT
+
+# %% [markdown]
+# **YOUR TASK:**
+# Create a plot that shows the k-means clusters next to an RGB image of the area. You may need to brighted your RGB image by multiplying it by 10.
+
+# %%
+# Unsupervised classification means we don't know what the groups are
+numbers = [1, 2, 3]
+groups = [f"Group {number}" for number in numbers]
+
+def plot_classified_data(processed_xr, prediction, groups):
+    # Plotting the classified data next to the original data
+    f, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 6))
+
+    ep.plot_rgb(processed_xr.values,
+                rgb=[0, 1, 2],
+                ax=ax1)
+    ax1.set_title("Composite RGB Image")
+
+    im2 = ax2.imshow(prediction,
+                     cmap="viridis",
+                     interpolation='none')
+    ax2.axis('off')
+    ep.draw_legend(im_ax=im2,
+                   titles=groups)
+    ax2.set_title("K-Means Classified Image")
+
+    plt.show()
+plot_classified_data(processed_xr, prediction, groups)
+
+# %%
+spectral_df = (
+    bands_stacked
+    .drop('spatial_ref')
+    .to_dataframe(name='reflectance')
+    .unstack('band')
+)
+# Drop the 'reflectance' level
+spectral_df.columns = spectral_df.columns.droplevel()
+spectral_df = spectral_df.reset_index(drop=True)
+spectral_df.columns = spectral_df.columns.astype(str)
+spectral_df.to_csv(os.path.join(cache_dir, 'spectral_df.csv'))
+
+plt.scatter(spectral_df['4'], spectral_df['5'])
```

Comment:
spectral_df is a spot where it would be nice to see some of those band labels you created up top. Another advantage of a dictionary is it can also be used to rename DataFrame columns.